pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+mod types;
+
+mod curves_parameters;
+#[frame_support::pallet]
+pub mod pallet {
+
+	use frame_support::{
+		dispatch::DispatchResult,
+		pallet_prelude::*,
+		traits::{
+			fungible::{Inspect as InspectFungible, MutateHold},
+			fungibles::{
+				metadata::{Inspect as FungiblesInspect, Mutate as FungiblesMetadata},
+				Create as CreateFungibles, Destroy as DestroyFungibles, Inspect as InspectFungibles,
+				Mutate as MutateFungibles,
+			},
+			AccountTouch,
+		},
+		Parameter,
+	};
+	use frame_system::pallet_prelude::*;
+	use parity_scale_codec::FullCodec;
+	use scale_info::TypeInfo;
+	use sp_runtime::{
+		traits::{CheckedAdd, CheckedSub, One, Saturating, Zero},
+		BoundedVec, FixedPointNumber,
+	};
+	use sp_std::default::Default;
+
+	use crate::types::PoolDetails;
+
+	type AccountIdLookupOf<T> = <<T as frame_system::Config>::Lookup as sp_runtime::traits::StaticLookup>::Source;
+
+	type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+
+	pub(crate) type DepositCurrencyBalanceOf<T> =
+		<<T as Config>::DepositCurrency as InspectFungible<<T as frame_system::Config>::AccountId>>::Balance;
+
+	pub(crate) type CollateralCurrencyBalanceOf<T> =
+		<<T as Config>::CollateralCurrency as InspectFungibles<<T as frame_system::Config>::AccountId>>::Balance;
+
+	pub(crate) type FungiblesBalanceOf<T> =
+		<<T as Config>::Fungibles as InspectFungibles<<T as frame_system::Config>::AccountId>>::Balance;
+
+	type FungiblesAssetIdOf<T> =
+		<<T as Config>::Fungibles as InspectFungibles<<T as frame_system::Config>::AccountId>>::AssetId;
+
+	type CollateralAssetIdOf<T> =
+		<<T as Config>::CollateralCurrency as InspectFungibles<<T as frame_system::Config>::AccountId>>::AssetId;
+
+	type BoundedCurrencyVec<T> = BoundedVec<FungiblesAssetIdOf<T>, <T as Config>::MaxCurrencies>;
+
+	pub(crate) type CurrencyNameOf<T> = BoundedVec<u8, <T as Config>::MaxStringLength>;
+
+	pub(crate) type CurrencySymbolOf<T> = BoundedVec<u8, <T as Config>::MaxStringLength>;
+
+	pub(crate) type CurveParameterTypeOf<T> = <T as Config>::CurveParameterType;
+
+	// TODO: change CurveParameterTypeOf.
+	pub(crate) type PoolDetailsOf<T> =
+		PoolDetails<<T as frame_system::Config>::AccountId, CurveParameterTypeOf<T>, BoundedCurrencyVec<T>>;
+
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		/// The currency used for storage deposits.
+		type DepositCurrency: MutateHold<Self::AccountId, Reason = Self::RuntimeHoldReason>;
+		/// The currency used as collateral for minting bonded tokens.
+		type CollateralCurrency: MutateFungibles<Self::AccountId>
```

Comment:
I think it would be better to call this `CollateralCurrencies`, since it's not a simple currency but rather a set thereof.

Reply:
For the first version, we decided to have only a single Collateral Currency. Ingo was saying once in the JF he wants multiple, but we wanted to introduce that later.

Reply:
Yeah but if you require `fungible`s, with a final s, the pallet has no idea about the runtime-provided value, so it should make no assumptions. Unless you replace `fungibles` with `fungible`, if that's implemented for any single entry in the fungible crate, then it would make sense. Otherwise I would not bake in Kilt-specific assumptions into a pallet that is supposed to be generic.

Reply:
forgot to rename this on my PR actually

Reply:
now called `CollateralCurrencies`